### PR TITLE
Update amp-bind docs

### LINKED ISSUE #7737: Update documentation for AMP.setState()
We should clearly explain all syntax options and nature of state update (shallow vs. recursive merge).

### PR DESCRIPTION
**Pull request by choumx**
Fixes #7737.

- Add more detail about `AMP.setState`
- Clarify element-specific attributes after #8033 

/to @bpaduch

### CODE REVIEW

File: extensions/amp-bind/amp-bind.md [hunk 1]
```diff
@@ -89,30 +98,25 @@ In this example, `myState.foo` will evaluate to `"bar"`.
...
 - All attribute bindings are sanitized for unsafe URL protocols (e.g. `javascript:`)
 
-### Bindable attributes
+#### Element-specific attributes
 
-For non-AMP elements, most attributes accepted by the [AMP Validator](https://validator.ampproject.org/) are bindable.
+Most attributes accepted by the [AMP Validator](https://validator.ampproject.org/), for AMP and non-AMP elements, are bindable.
 
-For AMP components, the height and width attributes are bindable along with the following specific attributes:
+There are a few special bindable attributes without non-bindable siblings:
```
Comment (bpaduch) [inline extensions/amp-bind/amp-bind.md]: This sentence confuses me.  So, does the table below show elements that are bindable but only for those attributes? Also, you say a "few" but there's two.  Can you clarify this area.  Thx

File: extensions/amp-bind/amp-bind.md [hunk 2]
```diff
@@ -89,30 +98,25 @@ In this example, `myState.foo` will evaluate to `"bar"`.
 - For security reasons, binding to `innerHTML` is disallowed
 - All attribute bindings are sanitized for unsafe URL protocols (e.g. `javascript:`)
 
-### Bindable attributes
+#### Element-specific attributes
 
-For non-AMP elements, most attributes accepted by the [AMP Validator](https://validator.ampproject.org/) are bindable.
+Most attributes accepted by the [AMP Validator](https://validator.ampproject.org/), for AMP and non-AMP elements, are bindable.
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: Which ones aren't?

Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: TBD in #8056. We probably won't allow mutation of things like SVG and other elements that weren't in the (now removed) `BindValidator` rule spec.

Removed the comma.

File: extensions/amp-bind/amp-bind.md (hunk 1 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: The table shows additional attributes that don't exist normally. E.g. `[slide]` is bindable but `slide` is not a valid attribute on `amp-carousel`.

Removed "a few". Let me know if you think there's a clearer way of stating this.

File: extensions/amp-bind/amp-bind.md [hunk 3]
```diff
@@ -73,10 +71,21 @@ The implicit state can be initialized at page load with custom JSON. For example
...
-In this example, `myState.foo` will evaluate to `"bar"`.
+[Expressions](#expressions) can reference state variables via dot syntax. In this example, `myState.foo` will evaluate to `"bar"`.
+
+
+##### AMP.setState()
+
+State can be mutated by the new `AMP.setState()` [action](../../spec/amp-actions-and-events.md).
 
-## Data binding
+- `AMP.setState()` performs a *shallow merge* of its arguments with the document state.
```
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Good idea. Added a link to `Object.assign`.

File: extensions/amp-bind/amp-bind.md [hunk 4]
```diff
@@ -73,10 +71,21 @@ The implicit state can be initialized at page load with custom JSON. For example
...
+- `AMP.setState()` can override data initialized by `amp-state`.
+
+### Binding
+
+A binding is a special attribute of the form `[property]` that links an element's property to an [expression](#expressions).
```
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Wanted to differentiate between the attribute that includes the brackets and property within the brackets.

File: extensions/amp-bind/amp-bind.md [hunk 5]
```diff
@@ -89,30 +98,25 @@ In this example, `myState.foo` will evaluate to `"bar"`.
...
 
 `amp-bind` expressions are JS-like with some important differences:
 
-- Expressions may only access the implicit JSON state
-- Expressions do not have access to globals like `window` or `document`
+- Expressions may only access the document [state](#state)
+- Expressions do **not** have access to globals like `window` or `document`
 - Only whitelisted functions are allowed
-- Custom functions and control flow statements (e.g. `for`, `if`) are disallowed
+- Custom functions, classes and control flow statements (e.g. `for`) are disallowed
```
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Oxford comma eh? Not used elsewhere so leaving as-is is more consistent.

File: extensions/amp-bind/amp-bind.md [hunk 6]
```diff
@@ -89,30 +98,25 @@ In this example, `myState.foo` will evaluate to `"bar"`.
...
-| amp-carousel | slide |
-| amp-img | src, srcset, alt |
-| amp-selector | selected |
-| amp-video | src, srcset, alt, controls, loop, poster |
-
-See [BindValidator](./0.1/bind-validator.js) for the canonical set of bindable
-elements and attributes.
+| Component | Attribute | Details |
+| --- | --- | --- |
+| amp-carousel[type=slides] | `[slide]` | The currently displayed slide index. |
```
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Good idea, added links to ABE.

File: extensions/amp-bind/amp-bind.md [hunk 7]
```diff
@@ -89,30 +98,25 @@ In this example, `myState.foo` will evaluate to `"bar"`.
...
 `amp-bind` expressions are JS-like with some important differences:
 
-- Expressions may only access the implicit JSON state
-- Expressions do not have access to globals like `window` or `document`
+- Expressions may only access the document [state](#state)
+- Expressions do **not** have access to globals like `window` or `document`
 - Only whitelisted functions are allowed
-- Custom functions and control flow statements (e.g. `for`, `if`) are disallowed
+- Custom functions, classes and control flow statements (e.g. `for`) are disallowed
 - Undefined variables, array-index-out-of-bounds return `null` instead of throwing errors
```
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Done.

File: extensions/amp-bind/amp-bind.md [hunk 8]
```diff
@@ -100,14 +100,14 @@ When the **state** changes, expressions are re-evaluated and the bound elements'
 
 #### Element-specific attributes
 
-Most attributes accepted by the [AMP Validator](https://validator.ampproject.org/), for AMP and non-AMP elements, are bindable.
+Most attributes accepted by the [AMP Validator](https://validator.ampproject.org/) for AMP and non-AMP elements, are bindable.
 
-There are a few special bindable attributes without non-bindable siblings:
+There are also special bindable attributes without non-bindable siblings:
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: s/siblings/counterparts ?

File: extensions/amp-bind/amp-bind.md (hunk 4 above)
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: ok

File: extensions/amp-bind/amp-bind.md (hunk 8 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Done.

**Pull request merged**